[iree][global] Control the demotion of ops

pashu123 opened the pull request:
Introduces `demote-only` flag in `demote-contraction-inputs-to-bf16` to control the demotion of ops. For e.g., if `demote-only=conv` only conv ops will be demoted.

hanhanW commented on compiler/src/iree/compiler/GlobalOptimization/DemoteContractionInputsToBF16.cpp:
```diff
@@ -21,12 +21,24 @@ namespace mlir::iree_compiler::GlobalOptimization {
+const std::string kDemoteAll = "all";
+const std::string kDemoteOnlyConv = "conv";
+const std::string kDemoteOnlyMatmul = "matmul";
+const std::string kDemoteNone = "none";
```
I think we can use enum to manage it better; we can get rid of magic strings. Can you add the `enum class` to Passes.h, and use `clEnumValN` in td file? Here is an example, see `Passes.h` and `Passes.td`: https://github.com/iree-org/iree/commit/e6d8aa70904f182881f388137708c07e5127483e

pashu123 replied:
Done, please review.

hanhanW commented on compiler/src/iree/compiler/GlobalOptimization/Passes.cpp:
```diff
@@ -120,8 +114,9 @@ void buildGlobalOptimizationPassPipeline(
-                         createDemoteContractionInputsToBF16Pass)
+      .addPass([&]() {
+        return createDemoteContractionInputsToBF16Pass(demoteOnly::None);
+      })
```
We will miss the control at `iree-compile` level. The options defined in `td` file are only available at `iree-opt` level. So we need a `llvm::cl::opt` flag. E.g.,

```cpp
static llvm::cl::opt<VectorPreProcStrategy> clPProcStrategy(
    "iree-llvmcpu-vector-pproc-strategy",
    llvm::cl::desc("Set the strategy for pre-processing Linalg operation "
                   "before vectorization:"),
    llvm::cl::values(
        clEnumValN(VectorPreProcStrategy::Peeling, "peel",
                   "Peel iterations from the vector dimensions so that they "
                   "become multiple of the vector length"),
        clEnumValN(
            VectorPreProcStrategy::Masking, "mask",
            " Compute vector dimensions assuming vector masking support. "
            "Vector sizes may be rounded up to the nearest power of two "
            "and out-of-bounds elements would be masked-out."),
        clEnumValN(
            VectorPreProcStrategy::None, "none",
            "Do not apply any vectorization pre-processing transformation.")));
```

hanhanW commented on compiler/src/iree/compiler/GlobalOptimization/Passes.h:
```diff
@@ -17,6 +17,10 @@
 
+// Used by the demoteContractionInputsToBF16 pass to determine which ops to
+// demote.
+enum class demoteOnly { All, Conv, Matmul, None };
```
style nit: should start with a capital letter; perhaps rename it to `DemotionOption`.

nit2: move it to where the pass is defined, i.e., line 57.

hanhanW commented on compiler/src/iree/compiler/GlobalOptimization/test/demote_contraction_inputs_to_bf16.mlir:
```diff
@@ -1,4 +1,4 @@
-// RUN: iree-opt --split-input-file -iree-global-opt-demote-contraction-inputs-to-bf16 %s | FileCheck %s
+// RUN: iree-opt --split-input-file -iree-global-opt-demote-contraction-inputs-to-bf16="demote-only=all" %s | FileCheck %s
```
Can you add other options? E.g.,

```
// RUN: iree-opt ... ="demote-only=matmul" %s | FileCheck %s -check-prefix=MATMUL-ONLY
```

We don't need to add all the lit tests. We can select few of them to test.

hanhanW commented on compiler/src/iree/compiler/GlobalOptimization/DemoteContractionInputsToBF16.cpp:
```diff
@@ -24,9 +24,18 @@ namespace {
+  using OpInterfaceRewritePattern<linalg::LinalgOp>::OpInterfaceRewritePattern;
+  explicit DemoteContractionInputsToBF16Pattern(
+      MLIRContext *ctx,
+      mlir::iree_compiler::GlobalOptimization::demoteOnly &option)
```
style nit: drop `mlir::iree_compiler::GlobalOptimization::`, it is already within the namespace. Same for other places

pashu123 committed:
[iree][global] Control the demotion of ops

Introduces demote-only flag in demote-contraction-inputs-to-bf16 to
control the demotion of ops. For e.g., if demote-only=conv only conv ops
will be demoted.

hanhanW commented on compiler/src/iree/compiler/GlobalOptimization/Passes.cpp:
```diff
@@ -51,11 +51,17 @@ static llvm::cl::opt<bool> clEnableFuseHorizontalContractions(
+                   "Only demote convolution ops."),
+        clEnumValN(DemotionOption::Matmul, "matmul", "Only demote matmul ops."),
+        clEnumValN(DemotionOption::None, "none", "Demote no contraction ops.")),
+    llvm::cl::init(DemotionOption::All));
```
hmm, I think the default value should be `None`? It is an experimental flag, so we don't want to it be on by default.

pashu123 replied:
Yes. Done.

hanhanW reviewed (commented):
Thanks, just one nit!
